### ISSUE TITLE
narrowing type resolves to unexpected type

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#6302]
### Discussed in https://github.com/microsoft/pyright/discussions/6301

<div type='discussions-op-text'>

<sup>Originally posted by **rchl** November  2, 2023</sup>
I have a home brewed Promise implementation that I've been using for a while and I've noticed that at some point Pyright started showing a new issue with it that wasn't reported before. The types in this implementation have developed to the point that I don't understand them myself anymore so I have to ask whether the current error is expected or not.

This is a very stripped down code that shows the error:

```py
from typing import Callable, Generic, List, Optional, Protocol, Tuple, TypeVar, Union

T = TypeVar('T')
T_contra = TypeVar('T_contra', contravariant=True)
TResult = TypeVar('TResult')


class ResolveFunc(Protocol[T_contra]):
    def __call__(self, resolve_value: T_contra) -> None:
        ...


FullfillFunc = Callable[[T], Union[TResult, 'Promise[TResult]']]


class Promise(Generic[T]):

    def then(self, onfullfilled: FullfillFunc[T, TResult]) -> 'Promise[TResult]':

        def callback_wrapper(resolve_fn: ResolveFunc[TResult], resolve_value: T) -> None:
            result = onfullfilled(resolve_value)
            if isinstance(result, Promise):
                result.then(lambda value: resolve_fn(value))
            else:
                resolve_fn(result)

                pass

        return Promise()
```

The error is:

```
  /usr/local/workspace/py-test/foo.py:23:54 - error: Argument of type "None" cannot be assigned to parameter "resolve_value" of type "TResult@then" in function "__call__"
    Type "None" cannot be assigned to type "TResult@then" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
```

This is line 23:
```
:23                result.then(lambda value: resolve_fn(value))
```</div>

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.335, which I just published. It will also be included in a future release of pylance.
